src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -0,0 +1,4787 @@
+
+	nLen = strlen(path);
+
+	p_Offset_Dots = strstr(path, "/./");
+
+	while (p_Offset_Dots > 0) {
+		p_Offset_Dots[0] = 0;
+		p_Offset_Dots[1] = 0;
+		p_Offset_Dots    = strstr(p_Offset_Dots + 2, "/./");
+		if (p_Offset_Dots == NULL)
+			break;
+	}
+
+	/* replace "//" with "/" */
+	p_Offset_Slash = strstr(path, "//");
+	while (p_Offset_Slash > 0) {
+		p_Offset_Slash[0] = 0;
+		p_Offset_Slash    = strstr(p_Offset_Slash + 1, "//");
+		if (p_Offset_Slash == NULL)
+			break;
+	}
+
+	/* remove '/.' at the end */
+	if (nLen > 2 && strncmp(path + nLen - 2, "/.", 3) == 0) {
+		p_Offset_Slash = path + nLen - 2;
+		p_Offset_Slash[0] = 0;
+		p_Offset_Slash[1] = 0;
+	}
+
+	new_str = path;
+	for (i = 0; i < nLen; i++) {
+		if (path[i]) {
+			new_str[nNonZero] = path[i];
+			nNonZero++;
+		}
+	}
+	/* remove "/" at the end of path */
+	new_str[nNonZero] = 0;
+	if (new_str[1] == 0 && new_str[0] == '/')
+		/* root dir */
+		return 1;
+	for (i = nNonZero - 1; i >= 0; i--) {
+		if (new_str[i] == '/') {
+			new_str[i] = 0;
+			nNonZero--;
+		} else {
+			break;
+		}
+	}
+
+	return nNonZero;
+}
+
+static void
+init_fd_list(void)
+{
+	if (pthread_mutex_init(&lock_fd, NULL) != 0) {
+		printf("\n mutex create_new_lock lock_fd init failed\n");
+		exit(1);
+	}
+	if (pthread_mutex_init(&lock_dirfd, NULL) != 0) {
+		printf("\n mutex create_new_lock lock_dirfd init failed\n");
+		exit(1);
+	}
+	if (pthread_mutex_init(&lock_mmap, NULL) != 0) {
+		printf("\n mutex create_new_lock lock_mmap init failed\n");
+		exit(1);
+	}
+
+	memset(file_list, 0, sizeof(struct file_obj *) * MAX_OPENED_FILE);
+	memset(dir_list, 0, sizeof(struct dir_obj *) * MAX_OPENED_DIR);
+	memset(mmap_list, 0, sizeof(struct mmap_obj) * MAX_MMAP_BLOCK);
+
+	next_free_fd    = 0;
+	last_fd         = -1;
+	next_free_dirfd = 0;
+	last_dirfd      = -1;
+	next_free_map   = 0;
+	last_map        = -1;
+	num_fd = num_dirfd = num_map = 0;
+}
+
+static int
+find_next_available_fd(struct file_obj *obj)
+{
+	int i, idx = -1, rc;
+	struct file_obj *new_obj = NULL;
+	bool allocated = false;
+
+	if (obj == NULL) {
+		D_ALLOC(new_obj, sizeof(struct file_obj));
+		if (new_obj == NULL) {
+			errno = ENOMEM;
+			return (-1);
+		}
+		new_obj->file      = NULL;
+		new_obj->idx_mmap  = -1;
+		new_obj->ref_count = 0;
+		allocated          = true;
+	} else {
+		new_obj = obj;
+	}
+
+	rc = pthread_mutex_lock(&lock_fd);
+	D_ASSERT(rc == 0);
+	if (next_free_fd < 0) {
+		rc = pthread_mutex_unlock(&lock_fd);
+		D_ASSERT(rc == 0);
+		if (allocated)
+			D_FREE(new_obj);
+		return next_free_fd;
+	}
+	idx = next_free_fd;
+	if (idx >= 0) {
+		new_obj->ref_count++;
+		file_list[idx] = new_obj;
+	}
+	if (next_free_fd > last_fd)
+		last_fd = next_free_fd;
+	next_free_fd = -1;
+
+	for (i = idx + 1; i < MAX_OPENED_FILE; i++) {
+		if (file_list[i] == NULL) {
+			/* available, then update next_free_fd */
+			next_free_fd = i;
+			break;
+		}
+	}
+	if (next_free_fd < 0)
+		printf("WARNING> All space for file_list are used.\n");
```

Comment:
What is the implication of this, should an error be returned here?

Reply:
Thank you! This is removed now. It was a warning about fd allocated reaching the limit. Now it just returns error when allocation fails due to out of space.